components/Heading/Heading.js

```diff
@@ -1,33 +1,188 @@
-import React from 'react';
+import React, { Component } from 'react';
 import { string, number, oneOfType, bool, node, oneOf } from 'prop-types';
 import classNames from 'classnames';
+import LinkIcon from 'static/images/icons/FontAwesome/link-solid.svg';
 import styles from './Heading.css';
 
-Heading.propTypes = {
-  className: string,
-  id: oneOfType([string, number]), // reference for scroll anchors
-  children: node.isRequired,
-  hasHeadingLines: bool,
-  theme: oneOf(['gray', 'secondary', 'white']),
-};
-
-Heading.defaultProps = {
-  className: undefined,
-  id: '',
-  hasHeadingLines: true,
-  theme: 'gray',
-};
-
-function Heading({ children, className, hasHeadingLines, id, theme }) {
-  const classes = classNames(className, styles.Heading, styles[theme], {
-    [`${styles.headingLines}`]: hasHeadingLines,
-  });
-
-  return (
-    <h2 className={classes} id={id}>
-      {children}
-    </h2>
-  );
+class Heading extends Component {
+  static propTypes = {
+    className: string,
+    id: oneOfType([string, number]), // reference for scroll anchors
+    anchorId: string,
+    customAnchorClass: string,
+    children: node.isRequired,
+    hasHeadingLines: bool,
+    hasHashLink: bool,
+    headingLevel: number,
+    theme: oneOf(['gray', 'secondary', 'white']),
+  };
+
+  static defaultProps = {
+    className: undefined,
+    id: '',
+    anchorId: 'default',
+    customAnchorClass: '',
+    hasHeadingLines: false,
+    hasHashLink: true,
+    headingLevel: 2,
+    theme: 'secondary',
+  };
+
+  state = {
+    isLinkIconVisible: false,
+  };
+
+  getAnchorId = () => {
+    const { props } = this;
+
+    return props.anchorId
+      .replace(/\s+/g, '-')
+      .replace(/\?|!|\./g, '')
+      .toLowerCase();
+  };
+
+  getVisibleIcon = () => {
+    const { props } = this;
+    const { theme } = props;
+    let visibleIcon = `${styles.icon} ${styles.iconVisible}`;
+
+    if (theme === 'white' || theme === 'gray') {
+      visibleIcon += ` ${styles.iconFillBlue}`;
+    } else {
+      visibleIcon += ` ${styles.iconFillWhite}`;
+    }
+
+    return visibleIcon;
+  };
+
+  toggleVisible = display => {
```

Comment:
The issue mentioned to make it visible on hover. Did you always want it visible and not just on hover?

Reply:
Sorry - it looks you do always have it rendered, but note that `visibility: hidden` also hides the content for screen readers. Note that there exist many web users that do not use a mouse - sometimes due to a disability. Since you only make it visible `onMouseEnter`, it would be considered "inaccessible".

Note the styles that we use for the `ScreenReaderOnly` component.

```css
.ScreenReaderOnly {
  border: 0;
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}
```

This will visibly hide the content, but still make it keyboard accessible. You could also use `onFocus` then for people who _can_ see, but choose to use the keyboard.

Reply:
Does this work?
```
<ScreenReaderOnly>
  {isLinkIconVisible ? 'Hashlink icon visible' : 'Hashlink icon hidden'}
</ScreenReaderOnly>
<LinkIcon className={isLinkIconVisible ? visibleIcon : hiddenIcon} />
```

Reply:
I didn't mean to imply you need to use `ScreenReaderOnly` component, but that you should copy it's styles for the `hiddenIcon` class you created.

Reply:
Also, there's a bonus round if you're interested!

I really don't care too much about this point, buuuuuuuuuuuuut...

Keep in mind there's a `:focus` and `:hover` in CSS, so you can do the visibility / invisibility with just CSS, making this component even simpler.

Reply:
Adding the ScreenReaderOnly style to the link hides it, but how do you bring it back on hover? `.hashLinkContainer a {
  /* visibility: hidden; */
  border: 0;
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.hashLinkContainer:hover a {
  visibility: visible;
}`